sdk/compiler/daml-lf-proto-decode/src/DA/Daml/LF/Proto3/DecodeV2.hs

```diff
@@ -783,25 +807,32 @@ decodeTypeLevelNat m =
 
 decodeType :: LF2.Type -> Decode Type
 decodeType LF2.Type{..} = mayDecode "typeSum" typeSum $ \case
-  LF2.TypeSumVar (LF2.Type_Var var args) ->
+  LF2.TypeSumVar (LF2.Type_Var var args) -> do
+    nullIfLfFlat_ args "Type_Var"
     decodeWithArgs args $ TVar <$> decodeNameId TypeVarName var
   LF2.TypeSumNat n -> TNat <$> decodeTypeLevelNat (fromIntegral n)
-  LF2.TypeSumCon (LF2.Type_Con mbCon args) ->
+  LF2.TypeSumCon (LF2.Type_Con mbCon args) -> do
+    nullIfLfFlat_ args "Type_Con"
     decodeWithArgs args $ TCon <$> mayDecode "type_ConTycon" mbCon decodeTypeConId
   LF2.TypeSumSyn (LF2.Type_Syn mbSyn args) ->
     TSynApp <$> mayDecode "type_SynTysyn" mbSyn decodeTypeSynId <*> traverse decodeType (V.toList args)
   LF2.TypeSumBuiltin (LF2.Type_Builtin (Proto.Enumerated (Right prim)) args) -> do
+    nullIfLfFlat_ args "Type_Builtin"
     decodeWithArgs args $ TBuiltin <$> decodeBuiltin prim
   LF2.TypeSumBuiltin (LF2.Type_Builtin (Proto.Enumerated (Left idx)) _args) ->
     throwError (UnknownEnum "Builtin" idx)
   LF2.TypeSumForall (LF2.Type_Forall binders mbBody) -> do
+    singletonIfLfFlat_ binders
     body <- mayDecode "type_ForAllBody" mbBody decodeType
     foldr TForall body <$> traverse decodeTypeVarWithKind (V.toList binders)
   LF2.TypeSumStruct (LF2.Type_Struct flds) ->
     TStruct <$> mapM (decodeFieldWithType FieldName) (V.toList flds)
   LF2.TypeSumInternedType n -> do
     DecodeEnv{internedTypes} <- ask
     lookupInterned internedTypes BadTypeId n
+  LF2.TypeSumTapp (LF2.Type_TApp lhs rhs) ->
+    TApp <$> mayDecode "type_TApp_lhs" lhs decodeType
```

Comment:
We should only decode this if running an LF version that supports flattening, since it didn't exist before.
```suggestion
    assertSupportsFlattening_
    TApp <$> mayDecode "type_TApp_lhs" lhs decodeType
```